### PR TITLE
Add drag support for dockwidgets sharing same position

### PR DESCRIPTION
**Pull request by goanpeca**
## Description

This related to dockwidgets that are tabified. Dockwidgets, when tabified, are put inside a QTabBar, but there is no access method to this bar, so to get it and process the event an eventFilter plus some search is needed. I added all the logic inside the event filter class.. but not sure if it is the best approach.

I attach a [video](http://youtu.be/wfvJupX4T5U) so you get the idea.

### CODE REVIEW

**Comment by Nodd** [comment]
Nice ! Is it you playing the guitar on the video ? :smile: 
What would you put in the contextual menu ? Entries for closing/setfloating/maximizing ? I think it would be helpful only it the titlebars are hidden as per https://github.com/spyder-ide/spyder/pull/2345#issuecomment-96438578

**Comment by goanpeca** [comment]
Hehe, no, the software I use puts the audio in, but i do play the guitar actually :-)

Now ... well I am considering two things.. if ALL widgets have a corner menu (most do on the upper right corner...) then ybe the options should go there... or also in the tab if right clicking...

I guess is natural to try to right click on a tab...

**Comment by Nodd** [comment]
It's natural to right click but I'd expect to see options about tab management, not about the content of the page. I don't know, maybe people would use it anyway...

**Comment by goanpeca** [comment]
Yes that is what I want, to put tab related stuff... move to right, move to right end..., move to left, move to left end..., close tab, float panel... etc

**Comment by Nodd** [comment]
Honnestly I'm not sure the menu is worth it. In FIrefox I regularely have more than 50 tabs opened and it's a mess to manage, but in spyder I have at most 5 dockpanels at the same place, so actions like move to left or move to end are somewhat useless.

**Comment by goanpeca** [comment]
I see what you mean, then this is almost ready :p

**Comment by ccordoba12** [comment]
First let me say that this is terrific addition!! Right now the way to reorganize dock widgets is to undock and dock them again, so this improves things by an order of magnitude!!

I also agree with @Nodd, I think adding a contextual menu is too much. Let's leave things simple for now :-)

**Comment by goanpeca** [comment]
Ok, I added some docstrings and is ready for a thorough revision ;-) @ccordoba12

**Comment by goanpeca** [comment]
@ccordoba12 this one is good to go, and easier to check (than the editor ine... that is still WIP)

File: spyderlib/plugins/__init__.py [hunk 1]
```diff
@@ -124,6 +126,14 @@ def __init__(self, *args, **kwargs):
         if sys.platform == 'darwin':
             self.setStyleSheet(self.DARWIN_STYLE)
 
+        # Needed for the installation of the event filter
+        self.title = args[0]
+        self.main = args[1]
```
Comment (ccordoba12) [inline spyderlib/plugins/__init__.py]: Where is `args` coming from?

File: spyderlib/plugins/__init__.py [hunk 2]
```diff
@@ -124,6 +126,14 @@ def __init__(self, *args, **kwargs):
         if sys.platform == 'darwin':
             self.setStyleSheet(self.DARWIN_STYLE)
 
+        # Needed for the installation of the event filter
+        self.title = args[0]
+        self.main = args[1]
+        self.bar = None
```
Comment (ccordoba12) [inline spyderlib/plugins/__init__.py]: wouldn't `dock_bar` be better, instead of `bar` here?

File: spyderlib/plugins/__init__.py [hunk 3]
```diff
@@ -131,6 +141,118 @@ def closeEvent(self, event):
...
+    def __init__(self, tabbar, main):
+        QObject.__init__(self)
+        self.tabbar = tabbar
+        self.main = main
+        self.moving = False
+        self.start_tab = None
+        self.end_tab = None
+
+    # Helper methods
+    def _get_widget(self, index):
```
Comment (ccordoba12) [inline spyderlib/plugins/__init__.py]: Why not call this `_get_plugin` instead?

File: spyderlib/plugins/__init__.py (hunk 1 above)
Comment (goanpeca) [inline spyderlib/plugins/__init__.py]: when create_dockwidget is called the args passed are the title and the parent (the main window in this case)

I agree is not so clear, but I am not so sure how to make it better...

File: spyderlib/plugins/__init__.py (hunk 3 above)
Comment (ccordoba12) [inline spyderlib/plugins/__init__.py]: I say it because I assume this is returning a `SpyderPlugin` instance, right?

File: spyderlib/plugins/__init__.py (hunk 2 above)
Comment (goanpeca) [inline spyderlib/plugins/__init__.py]: bar here refers to tabbar... so maybe I should change it

File: spyderlib/plugins/__init__.py (hunk 3 above)
Comment (goanpeca) [inline spyderlib/plugins/__init__.py]: Well.. we should change the variable in spyder.py that is called widgetlist (other PR) ;-) ... and yes I agree is better with plugins...

File: spyderlib/plugins/__init__.py (hunk 2 above)
Comment (ccordoba12) [inline spyderlib/plugins/__init__.py]: But it's the `tabbar` that holds dockwidgets, right?

So another possibility would be `dock_tabbar` :-)

File: spyderlib/plugins/__init__.py (hunk 1 above)
Comment (ccordoba12) [inline spyderlib/plugins/__init__.py]: But we have `get_plugin_title` and `self.main` already. I don't get it ;-)

File: spyderlib/plugins/__init__.py (hunk 2 above)
Comment (goanpeca) [inline spyderlib/plugins/__init__.py]: ok sure :)

File: spyderlib/plugins/__init__.py (hunk 3 above)
Comment (ccordoba12) [inline spyderlib/plugins/__init__.py]: Ok

File: spyderlib/plugins/__init__.py [hunk 4]
```diff
@@ -131,6 +141,118 @@ def closeEvent(self, event):
...
+            return True
+        if event_type == QEvent.MouseButtonRelease:
+            self.tab_released(event)
+            return True
+        return False
+
+    def tab_pressed(self, event):
+        """Method called when a tab from a QTabBar has been pressed."""
+        if event.button() == Qt.LeftButton:
+            self.start_tab = self.tabbar.tabAt(event.pos())
```
Comment (ccordoba12) [inline spyderlib/plugins/__init__.py]: Do you want to change the mouse icon here too?

Comment (ccordoba12) [inline spyderlib/plugins/__init__.py]: Ups, sorry, it's below :-)

File: spyderlib/plugins/__init__.py (hunk 1 above)
Comment (goanpeca) [inline spyderlib/plugins/__init__.py]: But those things belong to the plugin, not to the dockwidget,

File: spyderlib/plugins/__init__.py (hunk 4 above)
Comment (goanpeca) [inline spyderlib/plugins/__init__.py]: Yep. only when it starts to move ;-)

**Comment by goanpeca** [comment]
@ccordoba12 added corrections.

With regard to the question of the dockwidget... `SpyderPlugin` has a method create_dockwidget that uses `SpyderDockwidget` but the methods are from the plugin (SpyderPlugin) not from the widget, that is why I need to store those values when an instance if `SpyderDockwidget` is created...

File: spyderlib/plugins/__init__.py [hunk 5]
```diff
@@ -55,6 +57,97 @@ def get_icon(self):
...
+
+    def tab_pressed(self, event):
+        """Method called when a tab from a QTabBar has been pressed."""
+        if event.button() == Qt.LeftButton:
+            self.start_tab = self.dock_tabbar.tabAt(event.pos())
+
+    def tab_moved(self, event):
+        """Method called when a tab from a QTabBar has been moved."""
+        # If the left button isn't pressed anymore then return
+        if not event.buttons() & Qt.LeftButton:
```
Comment (ccordoba12) [inline spyderlib/plugins/__init__.py]: Why not a regular `and`, instead of `&` here?

File: spyderlib/plugins/__init__.py [hunk 6]
```diff
@@ -55,6 +57,97 @@ def get_icon(self):
...
+            self.move_tab(self.start_tab, self.end_tab)
+            self.start_tab = self.end_tab
+
+    def tab_released(self, event):
+        """Method called when a tab from a QTabBar has been released."""
+        QApplication.restoreOverrideCursor()
+        self.moving = False
+
+    def move_tab(self, start_tab, end_tab):
+        """Move a tab from a start to and end position ."""
```
Comment (ccordoba12) [inline spyderlib/plugins/__init__.py]: There's an unnecessary blank after `end position`

File: spyderlib/plugins/__init__.py [hunk 7]
```diff
@@ -124,6 +217,14 @@ def __init__(self, *args, **kwargs):
         if sys.platform == 'darwin':
             self.setStyleSheet(self.DARWIN_STYLE)
 
+        # Needed for the installation of the event filter
+        self.title = args[0]
+        self.main = args[1]
+        self.dock_tabbar = None
+
+        # Update the QTabBar everytime dockwidget visibility changes
+        self.visibilityChanged.connect(self.install_tab_event_filter)
```
Comment (ccordoba12) [inline spyderlib/plugins/__init__.py]: Why is this necessary?

File: spyderlib/plugins/__init__.py (hunk 1 above)
Comment (ccordoba12) [inline spyderlib/plugins/__init__.py]: I see. Then you need to redefine the __init__ method of `SpyderDockWidget` to accept the right args, i.e. something like

```python
def __init__(self, title, main, **kwargs):
    super(SpyderDockWidget, self).__init__(**kwargs)
    self.title = title
    self.main = main
    ...
```

File: spyderlib/plugins/__init__.py (hunk 5 above)
Comment (goanpeca) [inline spyderlib/plugins/__init__.py]: hmm True, proabably a leftover form the C++ code I took as example

File: spyderlib/plugins/__init__.py (hunk 6 above)
Comment (goanpeca) [inline spyderlib/plugins/__init__.py]: Got it

File: spyderlib/plugins/__init__.py (hunk 7 above)
Comment (goanpeca) [inline spyderlib/plugins/__init__.py]: When you move a dockwidget to a new position (location that was not existing before) this event gets triggered so I can add the filter to the QTabBar that gets created.

Was the only way I could think of so far...

File: spyderlib/plugins/__init__.py (hunk 5 above)
Comment (goanpeca) [inline spyderlib/plugins/__init__.py]: No I checked... buttons is an Or combination so I need the Binary And there...

http://doc.qt.io/qt-4.8/qmouseevent.html#buttons

File: spyderlib/plugins/__init__.py (hunk 1 above)
Comment (goanpeca) [inline spyderlib/plugins/__init__.py]: Fixed

File: spyderlib/plugins/__init__.py (hunk 7 above)
Comment (goanpeca) [inline spyderlib/plugins/__init__.py]: Also it even though this operation is triggered often... the filter is only installed once per QTabBar, so it should not impact usability. I hope there was a more intuitive and direct way offered by Qt, but I guess dockwidgets are meant to be used as extra panels in an application with a main  widget.. ;-)

Comment (goanpeca) [inline spyderlib/plugins/__init__.py]: Oops I should have used `dockLocationChanged`

Will adapt the coed

File: spyderlib/plugins/__init__.py [hunk 8]
```diff
@@ -55,6 +57,97 @@ def get_icon(self):
...
+            self.move_tab(self.start_tab, self.end_tab)
+            self.start_tab = self.end_tab
+
+    def tab_released(self, event):
+        """Method called when a tab from a QTabBar has been released."""
+        QApplication.restoreOverrideCursor()
+        self.moving = False
+
+    def move_tab(self, start_tab, end_tab):
+        """Move a tab from a start to and end position."""
```
Comment (ccordoba12) [inline spyderlib/plugins/__init__.py]: `to an end`, not `to and end` :-)

File: spyderlib/plugins/__init__.py [hunk 9]
```diff
@@ -55,6 +57,97 @@ def get_icon(self):
...
+        plugins = self._get_plugins()
+        start_plugin = self._get_plugin(start_tab)
+        end_plugin = self._get_plugin(end_tab)
+
+        start = plugins.index(start_plugin)
+        end = plugins.index(end_plugin)
```
Comment (ccordoba12) [inline spyderlib/plugins/__init__.py]: What about `start_idx` and `end_idx` for these variable names? I think they should be clearer ;-)

File: spyderlib/plugins/__init__.py [hunk 10]
```diff
@@ -119,11 +212,21 @@ class SpyderDockWidget(QDockWidget):
 
     plugin_closed = Signal()
 
-    def __init__(self, *args, **kwargs):
-        super(SpyderDockWidget, self).__init__(*args, **kwargs)
+    def __init__(self, title, parent):
+        super(SpyderDockWidget, self).__init__(title, parent)
```
Comment (ccordoba12) [inline spyderlib/plugins/__init__.py]: Have you tried this in PyQt5? Is it working correctly there?

File: spyderlib/plugins/__init__.py (hunk 9 above)
Comment (goanpeca) [inline spyderlib/plugins/__init__.py]: Sure I can change...

File: spyderlib/plugins/__init__.py (hunk 10 above)
Comment (goanpeca) [inline spyderlib/plugins/__init__.py]: Not really, but I do not see why it would not work with PyQt5

Comment (goanpeca) [inline spyderlib/plugins/__init__.py]: I will check later

Comment (ccordoba12) [inline spyderlib/plugins/__init__.py]: Sorry, problems with PyQt5 only happen on mixins. So, there shouldn't be any problem with this one :-)

**Comment by goanpeca** [comment]
I renamed variables to be more consistent with the other PR (editor draggable tabs) and added support to call context menus if needed. For the moment the methods are empty, but I added them in case they are useful in the future.

I also included the cursor fix so that it compensates when the moved tabs have different size.

I made suggested corrections and everything now is working as expected :-), so if @ccordoba12 can check and approve, this one is ready.

File: spyderlib/plugins/__init__.py [hunk 11]
```diff
@@ -55,6 +57,136 @@ def get_icon(self):
...
+    # Helper methods
+    def _get_plugin(self, index):
+        """Get plugin reference based on tab index."""
+        for plugin in self.main.widgetlist:
+            if plugin.get_plugin_title() == self.dock_tabbar.tabText(index):
+                return plugin
+
+    def _get_plugins(self):
+        """
+        Get a list of all the plugins references in the QTabBar to which this
```
Comment (ccordoba12) [inline spyderlib/plugins/__init__.py]: Please change this to `Get a list of all plugin references in the QTabbar ...`

Comment (goanpeca) [inline spyderlib/plugins/__init__.py]: Sure :+1:

File: spyderlib/plugins/__init__.py [hunk 12]
```diff
@@ -55,6 +57,136 @@ def get_icon(self):
...
+        for i in range(len(plugins)-1):
+            self.main.tabify_plugins(plugins[i], plugins[i+1])
+        from_plugin.dockwidget.raise_()
+
+    def show_tab_menu(self):
+        """Show the context menu assigned to tabs."""
+
+    def show_nontab_menu(self):
+        """Show the context menu assigned to nontabs section."""
+
```
Comment (ccordoba12) [inline spyderlib/plugins/__init__.py]: Aren't these two last methods missing a `pass`

Comment (goanpeca) [inline spyderlib/plugins/__init__.py]: I could add them, but the docstring replaces the need for a `pass`... I can add... or leave it like it is

Comment (ccordoba12) [inline spyderlib/plugins/__init__.py]: Please add `pass` to them. It feels weird without it :-)

File: spyderlib/plugins/__init__.py [hunk 13]
```diff
@@ -55,6 +57,136 @@ def get_icon(self):
...
+        if x < tab_x_min or x > tab_x_max:
+            new_pos = self.dock_tabbar.mapToGlobal(QPoint(x + delta, y))
+            cursor.setPos(new_pos)
+
+    def eventFilter(self,  obj,  event):
+        """Filter mouse press events."""
+        event_type = event.type()
+        if event_type == QEvent.MouseButtonPress:
+            self.tab_pressed(event)
+            return False
```
Comment (ccordoba12) [inline spyderlib/plugins/__init__.py]: Why these return values are needed? Could you add a comment about their need and why sometimes are True and others False?

File: spyderlib/plugins/__init__.py [hunk 14]
```diff
@@ -55,6 +57,136 @@ def get_icon(self):
...
+        if not self.moving and self.from_index != -1 and self.to_index != -1:
+            QApplication.setOverrideCursor(Qt.ClosedHandCursor)
+            self.moving = True
+
+        if self.to_index == -1:
+            self.to_index = self.from_index
+
+        from_index, to_index = self.from_index, self.to_index
+        if from_index != to_index and from_index != -1 and to_index != -1:
+            self.sig_tab_moved.emit(from_index, to_index)
```
Comment (ccordoba12) [inline spyderlib/plugins/__init__.py]: Why is this signal needed if nothing connects to it?

Comment (goanpeca) [inline spyderlib/plugins/__init__.py]: At this moment it does nothing, but I guess it could be useful if some behavior wants to be executed whenever a the tabs are reorganized. It is for completeness...

File: spyderlib/plugins/__init__.py [hunk 15]
```diff
@@ -119,11 +251,21 @@ class SpyderDockWidget(QDockWidget):
...
 
+        # Needed for the installation of the event filter
+        self.title = title
+        self.main = parent
+        self.dock_tabbar = None
+
+        # To track dockwidget changes the filter is installed when dockwidget
+        # visibility changes. This installs the filter on startup and also
+        # on dockwidgets that are undocked and then docked to a new location.
+        self.visibilityChanged.connect(self.install_tab_event_filter)
```
Comment (ccordoba12) [inline spyderlib/plugins/__init__.py]: Hadn't you mentioned that you were going to use a different signal here?

File: spyderlib/plugins/__init__.py (hunk 13 above)
Comment (goanpeca) [inline spyderlib/plugins/__init__.py]: An event filter intercepts the fired events, and it has to return True, if it propagates the events... or False if it captures the event.... 

It is the normal behavior ....

http://doc.qt.io/qt-4.8/eventsandfilters.html#event-filters

File: spyderlib/plugins/__init__.py (hunk 12 above)
Comment (goanpeca) [inline spyderlib/plugins/__init__.py]: Done

File: spyderlib/plugins/__init__.py (hunk 14 above)
Comment (ccordoba12) [inline spyderlib/plugins/__init__.py]: Then please remove it :-)

When (if) it's needed, it can easily be added. Right now it's just noise and adds confusion to the code.

File: spyderlib/plugins/__init__.py (hunk 13 above)
Comment (goanpeca) [inline spyderlib/plugins/__init__.py]: I will add some info to the docsting though

File: spyderlib/plugins/__init__.py (hunk 15 above)
Comment (goanpeca) [inline spyderlib/plugins/__init__.py]: I did, but then I remembered that it did not work properly, and I revert to this one. The problem of using the other signal is that it does not work when a dockwidget is moved to a new position.

File: spyderlib/plugins/__init__.py (hunk 14 above)
Comment (goanpeca) [inline spyderlib/plugins/__init__.py]: Done

File: spyderlib/plugins/__init__.py (hunk 15 above)
Comment (ccordoba12) [inline spyderlib/plugins/__init__.py]: Ok, no problem.

File: spyderlib/plugins/__init__.py (hunk 5 above)
Comment (ccordoba12) [inline spyderlib/plugins/__init__.py]: Ok, thanks.

File: spyderlib/plugins/__init__.py (hunk 13 above)
Comment (ccordoba12) [inline spyderlib/plugins/__init__.py]: Ok, thanks!

**Comment by ccordoba12** [comment]
Nothing more from me to say :-) Please make my suggested changes so I can start testing.

**Comment by goanpeca** [comment]
@ccordoba12 done

File: spyderlib/plugins/__init__.py [hunk 16]
```diff
@@ -55,6 +57,141 @@ def get_icon(self):
...
+            self.main.tabify_plugins(plugins[i], plugins[i+1])
+        from_plugin.dockwidget.raise_()
+
+    def show_tab_menu(self):
+        """Show the context menu assigned to tabs."""
+        pass
+
+    def show_nontab_menu(self):
+        """Show the context menu assigned to nontabs section."""
+        pass
```
Comment (ccordoba12) [inline spyderlib/plugins/__init__.py]: Las thing, I think I have the perfect suggestion for this entry

```python
    self.main.createPopupMenu()
```

Comment (goanpeca) [inline spyderlib/plugins/__init__.py]: Ehh ? what does it do?

Comment (goanpeca) [inline spyderlib/plugins/__init__.py]: http://doc.qt.io/qt-4.8/qmainwindow.html#createPopupMenu ?

Comment (ccordoba12) [inline spyderlib/plugins/__init__.py]: It shows the same context menu as when you do a right click over a toolbar.

Essentially, It shows the Help menu but with just a few entries.

Comment (goanpeca) [inline spyderlib/plugins/__init__.py]: it does not work...

Comment (goanpeca) [inline spyderlib/plugins/__init__.py]: Ok fixed it :), now this will only popup when clicking in an area where no tabs are present...

In this image I right clicked to the right of the `File explorer` Tab

image

**Comment by goanpeca** [comment]
@ccordoba12 you want to get the same context menu when right clicking on a tab as well??

I have two methods in this PR, one is `show_tab_menu` (right click on a tab) and `show_nontab_menu` (right click on a QTabBar empty space, generally to the right of the tabs)

**Comment by ccordoba12** [comment]
I'd say yes, for now. But leave the two methods, so if we come up with a better context menu for tabs on the future, we can change it easily :-)

**Comment by goanpeca** [comment]
Ready for testing @ccordoba12

**Comment by ccordoba12** [comment]
Working great, even in PyQt5!

Nothing else than merging ;-)
